src/lib.rs

```diff
@@ -145,6 +145,37 @@ pub trait CustomizeConnection<C, E>: fmt::Debug + Send + Sync + 'static {
     /// The default implementation does nothing.
     #[allow(unused_variables)]
     fn on_release(&self, conn: C) {}
+
+    /// Called with connections immediately after they are checked out from
+    /// the connection pool for later usage.
+    ///
+    /// This method can be used to customize the behaviour of `ConnectionManager::is_valid`
+    /// for specific usages like testing or other manual preperations.
+    ///
+    /// The default implementations simply returns the existing `is_valid_result`
+    ///
+    /// # Errors
+    ///
+    /// If this method returns an error, the connection will be discarded
+    #[allow(unused_variables)]
+    fn on_checkout(&self, conn: &mut C, is_valid_result: Result<(), E>) -> Result<(), E> {
```

Comment:
What are the use cases for wanting to interact with is_valid_result here and has_broken below?

Reply:
I cannot comment on general use cases here, as that is highly dependent on the `ConnectionManager` implementation and the underlying connections. Therefore I will only comment on diesel related use cases here.

For `on_checkin` I see two use cases:
* Allow to checkin transactions with open transactions into the pool, which is sometimes useful for tests. (Diesel marks connections with open transactions as broken)
* Trying to close open transactions and "unbreak" connections here

For `on_checkout` the main use case would be setting up things before checking out the connection, for example opening a dangling transaction. I've included the result of `is_valid` here for consistency with `on_checkin`, but I cannot present a diesel related use case to actually use this result. I would guess that there could be a theoretical connection implementation that allows to reconnect the internal connection if some check has failed.

Reply:
@sfackler Anything that's missing here?